### PR TITLE
Toggle Buttons for the User Table

### LINKED ISSUE #12: #3 | B1 | Show user roles in profile debugging information | PRIORITY
When you toggle the role using the Swagger endpoint, the roles `ROLE_ADMIN` and `ROLE_INSTRUCTOR` show up (or not) on the user profile in the debugging information under roles and in the rolesList as shown below (the example only shows `ROLE_ADMIN` appearing.)

image

### LINKED ISSUE #25: #3 | F1 | Add toggle buttons to UsersTable | PRIORITY
- [x] The UsersTable (currently used on `AdminUsersPage` and the `ProfilePage` has a new property `showToggleButtons` that defaults to false
- [x] When the showButtons property is true, the UsersTable component includes buttons to toggle admin and instructor status, similar to the buttons on the [UsersTable from proj-gauchoride](https://github.com/ucsb-cs156/proj-gauchoride/blob/main/frontend/src/main/components/Users/UsersTable.js)
- [x] A storybook story for UsersTable is added with two entries: one that shows the table without the buttons, and another with the buttons. ([This file](https://github.com/ucsb-cs156/proj-gauchoride/blob/main/frontend/src/stories/components/Users/UsersTable.stories.js) may be helpful as an example.)
- [x] The value of showToggleButtons on the AdminUsersPage is set to true, and the buttons there actually toggle the statuses.  
- [ ] Status changes are reflected both in the admin users table and on the actual users logged in session.

### PR DESCRIPTION
**Pull request by chauAlex**
This PR introduces toggle buttons for the User Table when the showToggleButtons property is true, and gives the ability to give and remove instructor and admin privileges. The AdminUsersPage has it set to true and thus shows the buttons, whereas the Profile Page does not provide a value for the showToggleButtons and thus defaults to false. It also reloads the page to display the current boolean values on the page.

Note as this branch is made aside from the interceptors, it requires a reload to show the ROLES in the user debugging info.

Screen Shot 2023-11-28 at 3 11 20 PM
Screen Shot 2023-11-28 at 3 11 36 PM

Link to PR storybook: https://ucsb-cs156-f23.github.io/proj-organic-f23-6pm-1/prs/59/storybook/?path=/story/components-users-userstable--admin-view-three-users

Dokku deployment for this branch:  https://proj-organic-chaualex.dokku-05.cs.ucsb.edu

Closes #25, Closes #12

### CODE REVIEW

**Review by Dagin-Gregory — DISMISSED** [review]
Looks good, I did get a 405 when trying to run toggle admin and instructor on the storybook, but I suspect that's due to there being no backend to support the storybook. (Unable to test on a dokku deployment)

**Review by Johnsonchan105 — DISMISSED** [review]
LGTM

**Review by pconrad — CHANGES_REQUESTED** [review]
Code LGTM.  Could you set up a dokku deployment for this branch and link to it in the PR description?

**Comment by chauAlex** [comment]
Sure thing, dokku deployment is now up

**Review by Joseph-Foster — DISMISSED** [review]
LGTM

**Comment by chauAlex** [comment]
Note: Once Johnson's branch is merged it should also finish #12

File: frontend/src/main/components/Users/UsersTable.js [hunk 1]
```diff
@@ -1,44 +1,99 @@
...
-export default function UsersTable({ users }) {
+    // Stryker disable all : hard to test for query caching
+    const toggleAdminMutation = useBackendMutation(
+        cellToAxiosParamsToggleAdmin,
+        {onSuccess: () => {
+            window.location.reload();
+        }},
+        ["/api/admin/users"]
+    );
+    // Stryker enable all
```
Comment (pconrad) [inline frontend/src/main/components/Users/UsersTable.js]: Should be `// Stryker restore all`

That's true throughout the code base.

**Review by pconrad — CHANGES_REQUESTED** [review]
So, there is one thing that you must fix:
* `// Stryker enable all` is incorrect (notwithstanding the many examples in the code, all of which eventually need to be corrected.  The correct directive is `// Stryker restore all`

Then there is another that will greatly improve the User Experience. Right now, toggling a status causes a full page refresh. But that's unnecessary.   The way to get the page to update without forcing a refresh is this:

Replace this line of code in `AdminUsersPage.js`

```js
const { data: users } = useUsers();
```

With a call to `useBackend`, like this:

```js
 const { data: users, error: _error, status: _status } =
        useBackend(
            // Stryker disable next-line all : don't test internal caching of React Query
            ["/api/admin/users"],
            // Stryker disable next-line StringLiteral,ObjectLiteral : since "GET" is default, "" is an equivalent mutation
            { method: "GET", url: "/api/admin/users" },
            // Stryker disable next-line all
            []
        );
```   


The reason this is important is that `useUsers` (which should not have been written in the first place) doesn't use React Query Library properly; in particular it doesn't set a caching key for retrieving the users.

You are correctly attempting to invalidate the cache key in your POST calls when you toggle the status:
```js
  const toggleAdminMutation = useBackendMutation(
        cellToAxiosParamsToggleAdmin,
        {onSuccess: () => {
            window.location.reload();
        }},
        ["/api/admin/users"]
    );
```

But for that to work, the  cache key you invalidate,  `["/api/admin/users"]` in the POST, has to match the cache key on the GET.   useBackend does this properly.  useUsers doesn't.

If you make this change, then it should work without having to force a page refresh.

And, you'll be able to remove the file where useUsers is defined, and where it is tested.

That part is optional, but the other three organic groups did it, so your app will be stand out (in a bad way) if you don't make this change.  It's not required, but it will make your app look better on demo day, which does impact your grade to an extent.

**Comment by ErwanDev** [comment]
All instances of // Stryker enable all have been replaced with // Stryker restore all. Made the changes needed to remove the unneeded page refresh and deleted useUsers files.

**Review by Joseph-Foster — APPROVED** [review]
LGTM, doesn't refresh page now

**Review by pconrad — APPROVED** [review]

**Review by pconrad — APPROVED** [review]
LGTM!

**Pull request merged**